### ISSUE TITLE
Sidebar boxes with autogenerated indexes

### ISSUE THREAD

**Issue by tristanlatr** [twisted/pydoctor#257]
Hello,

I'm using `pydoctor` as the only tool to document a little library https://mfesiem.github.io/docs/msiempy/msiempy.html

As you can see, there are numerous titles, and it would be great if users could be navigate rapidly to one another. 

In this issue I'm proposing to add a sidebar (it could be an option or enabled by default), just like in JavaDoc (i.e. https://docs.oracle.com/javase/7/docs/api/index.html?overview-summary.html) to all pages. 

JavaDoc do not create summary pages for modules, PyDoctor does, but it doesn't change anything...

Several level (boxes) of indexes can be identified:

1/ The table of contents. The result of ``.. contents::``. Would be present in module and classes documentation if titles are defined in the documented element's docstring. 
2/ The Classes / Interfaces / Exceptions / Constants / Functions / etc listed in the current module (As an easy way to navigate to related objects). 
3/ The module index 
Those 3 indexes could be in 3 ajustable boxes (like JavaDoc, again). 

I think this feature would greatly increase `pydoctor` usability and adoption.

**Comment by tristanlatr** [comment]
Here a really simple Frame view I got with

```
<frameset cols="30%,70%" title="Documentation frame">
    <frameset rows="30%,70%" title="Left frames">
    <frame src="moduleIndex.html" name="moduleIndex" title="All modules">
    <frame src="classIndex.html" name="classFrame" title="All classes">
    </frameset>
    <frame src="index.html" name="indexFrame" title="Index" scrolling="yes">
</frameset>
``` 

![by default 2020-11-06 at 9 26 37 PM](https://user-images.githubusercontent.com/19967168/98429775-c21f6800-2076-11eb-9a5a-e45f999d84e5.png)

It's only to show the type of thing that can be done. The module index and the class index files should not be used in the frames, but some other specially generated files, also it should not show the docstrings. The class list should only show classes in the current module (everything in `index.html`). The headers should be removed and the titles made smaller. The table of contents is not there, maybe it could be added on the top-right corner or another frame. 

I guess we'de have to edit the templates: `common.html` and `index.html`. 

My question are: Where do I start ? I'm not very familiar with introspection. Does a custom HTMLWriter is sufficient to do the job ?

**Comment by mthuurne** [comment]
The way pydoctor works is that information from parsing the source (AST = abstract syntax tree) and parsing the docstrings within that source is combined into a model and in a separate phase information from that model is put into an HTML report. So assuming that the model already contains all the information you need (I think it does), you only need to change the output generation, not the parsing/analysis code.

pydoctor uses [Twisted's templating system](https://twistedmatrix.com/documents/current/web/howto/twisted-templates.html), with the templates in the `pydoctor/templates/` directory and the code that fills in gaps in those templates in `pydoctor/templatewriter/pages/`.

I agree that we can make more efficient use of large screens. I'm not a fan of the JavaDoc layout though, it's quite busy and a bit cramped. We should design a layout that will only show scroll bars or cut off information in exceptional cases. And we should use responsive CSS rather than frames, such that the sidebar is either hidden (if redundant) or moved into the center part of the document (if essential) on narrow displays.

In my opinion, the class hierarchy isn't useful enough to dedicate sidebar space to. Modern OO design prefers composition over inheritance, so deep class hierarchies should be rare. And when they do exist, they are hard to present well in a small space.

Having the module or class index in a sidebar would be very useful. This could be done using only the name column + background color from the usual index table. Perhaps the summary string could be shown in tooltip style on hover.

One thing I've been wondering about is this element at the top of the page:

![image](https://user-images.githubusercontent.com/246676/98432487-09c9e200-20bf-11eb-80c8-b82a6424db34.png)

The levels in the namespace (`msiempy`, `core` and `session`) are clickable and I use this very often to navigate the documentation. However, the full name of the documented module/class contains the same namespace and is not clickable, even though it would make a good click target since it's much larger.

We could move the "View Source" and "View In Hierarchy" links to the part of the page that also displays the docstring, make the big namespace clickable and remove the then-redundant "Part of" element. This would result in a single line that acts both as a document title and as a navigation element, which we can then make sticky, meaning that it stays at the top of the screen when the document is scrolled.

With higher namespace levels reachable via the top bar, the sidebar could be used exclusively for navigating within the current namespace level (module or class).

**Comment by mthuurne** [comment]
I implemented the sticky navigation bar idea in #265.

**Comment by tristanlatr** [comment]
> You only need to change the output generation, not the parsing/analysis code.

I'll se what I can do and report back... 

> we should use responsive CSS rather than frames

I agree frames should be avoided. They are deprecated in HTML5

> With higher namespace levels reachable via the top bar, the sidebar could be used exclusively for navigating within the current namespace level (module or class).

I'm not sure on this one. There are several levels of indexes that can be discussed. 

Regarding index 2/
As an exemple, see: https://pawamoy.github.io/mkdocstrings/reference/handlers/python/ or https://timothycrosley.github.io/portray/reference/portray/exceptions/

I want to draw you attention to the right sidebar: It's shows the current module classes as well as methods and attributes. 
In `pydoctor`, we could not show other classes methods but only the currently selected one (if reviewing a class documentation), but still show other classes names as the quicker way to navigate to a class in the same module (One click instead of 2 using the top navbar)

Regarding index 3/
In both exemples, the module index is separated from the class/methods index, but may be it could be merged. 
They could be on the left side.

Regarding index 1/ 
It should show up only if the user defined some titles in the module or class docstring. 
It could be on the right side.

> In my opinion, the class hierarchy isn't useful enough to dedicate sidebar space to

I agree, it would be just a list of classes names (like in Javadoc)

> I implemented the sticky navigation bar idea in #265.

So nice! I'll check it out and give feedback.

**Comment by tristanlatr** [comment]
>  I'll check it out and give feedback.

I won't have the time to look at this in the near future actually.
